File: frontend/client/lib/models/poi.dart
import 'narration.dart';
import 'restaurant.dart';
import 'geofence.dart';

class POI {
  final int poiId;
  final String name;
  final String? imageUrl;
  final double latitude;
  final double longitude;
  final List<String> imageUrls;
  final List<Narration> narrations;
  final List<Restaurant> restaurants;
  final Geofence? geofence;

  POI({
    required this.poiId,
    required this.name,
    this.imageUrl,
    required this.latitude,
    required this.longitude,
    required this.imageUrls,
    this.narrations = const [],
    this.restaurants = const [],
    this.geofence,
  });

  factory POI.fromJson(Map<String, dynamic> json) {
    return POI(
      poiId: json['poiId'],
      name: json['name'],
      imageUrl: json['imageUrl'], // ✅ THÊM
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      imageUrls: List<String>.from(json['imageUrls'] ?? []),
      narrations: (json['narrations'] as List<dynamic>?)
              ?.map((n) => Narration.fromJson(n))
              .toList() ??
          [],
      restaurants: (json['restaurants'] as List<dynamic>?)
              ?.map((r) => Restaurant.fromJson(r))
              .toList() ??
          [],
      geofence:
          json['geofence'] != null ? Geofence.fromJson(json['geofence']) : null,
    );
  }
}

File: frontend/client/lib/model/POI.cs
using CommunityToolkit.Mvvm.ComponentModel;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Serialization;

namespace client.lib.model
{
    public partial class POI : ObservableObject
    {
        [JsonPropertyName("poiId")]
        public int PoiId { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; } = string.Empty;

        [JsonPropertyName("description")]
        public string? Description { get; set; }

        [JsonPropertyName("latitude")]
        public double Latitude { get; set; }

        [JsonPropertyName("longitude")]
        public double Longitude { get; set; }
        [JsonPropertyName("averageRating")]
        public double AverageRating { get; set; }

        [JsonPropertyName("reviewCount")]
        public int ReviewCount { get; set; }

        // Danh sách link ảnh từ API Cloudinary trả về
        [JsonPropertyName("imageUrls")]
        public List<string> ImageUrls { get; set; } = new();
        public List<Review> Reviews { get; set; } = new();

        [JsonIgnore]
        public string ImageUrl
        {
            get
            {
                if (ImageUrls == null || !ImageUrls.Any())
                    return "placeholder_img.webp"; 

                return ImageUrls.First();
            }
        }

        [JsonIgnore]
        public List<string> FullImageUrls
        {
            get
            {
                // Nếu không có ảnh nào, trả về 1 list chứa ảnh mặc định để UI không bị trống
                if (ImageUrls == null || !ImageUrls.Any())
                    return new List<string> { "placeholder_img.webp" };

                return ImageUrls;
            }
        }

        [ObservableProperty]
        private string _distanceDisplay = "Đang tính...";

        [ObservableProperty]
        private double _distanceInMeters;

        [JsonPropertyName("narrations")]
        public List<Narration> Narrations { get; set; } = new();

        [JsonPropertyName("restaurants")]
        public List<Restaurant> Restaurants { get; set; } = new();
    }
}